lvm: Expose logical volume substructure

mvollmer opened the pull request:
I think we need all this detail in the upper levels.  For example, to determine whether a top-LV is degraded.

blivetghoul commented:
Can one of the admins verify this patch?

tbzatek commented:
Jenkins, ok to test.

tbzatek commented:
No objections from my side. Would be interesting to see how is this visualized in upper layers.

mvollmer commented:
> No objections from my side. Would be interesting to see how is this visualized in upper layers.

Here is a quick demo of what I have in mind: https://www.youtube.com/watch?v=8xrgj7z-7-Y

mvollmer commented on src/lib/plugin_apis/lvm.api:
```diff
@@ -270,6 +271,94 @@ GType bd_lvm_vgdata_get_type () {
+    guint64 size_pe;
+    guint64 lv_start_pe;
+    guint64 pv_start_pe;
+    gchar *pvdev;  // NULL if missing
```
Good idea, I'll check this out. (We can certainly get the UUID for a present PV, but I am afraid we can't for a missing one...)

mvollmer replied:
No, we can't get the UUID. lvmdbusd does not provide any information about segments that have a missing PV.

mvollmer commented on src/lib/plugin_apis/lvm.api:
```diff
@@ -1242,6 +1343,20 @@ BDLVMLVdata* bd_lvm_lvinfo (const gchar *vg_name, const gchar *lv_name, GError *
+ * @vg_name: (nullable): name of the VG to get information about LVs from
+ * @error: (out) (optional): place to store error (if any)
+ *
+ * This function will fill out the data_lvs, metadata_lvs, and segs fields as well.
```
Yes, I am afraid that filling out the lists will change the space complexity of the result structure too much.  Existing callers would pay dearly for stuff that they don't need. I am hoping that between us we can figure out what is best.

vojtechtrefny replied:
Ok, if it would be too slow to get these results, then it makes sense to not get it in `bd_lvm_lvinfo`, but I think it might be too confusing (at least it confused me) to have two different results for two different calls to get `BDLVMLVdata`, I would prefer a separate struct for these "extra" parameters (it's too many structs already, but I think it would be better that way). Thoughts @tbzatek?

mvollmer replied:
I don't think it would be too slow: We already iterate over all segments in both bd_lvm_lvinfo and bd_lvm_lvs (and use a O(n^2) linear search to find duplicates, urks). The "tree" variant gets a couple more fields from `lvs`, but that should be acceptable (unless you have a million stripes and the "devices" property etc gets really huge).

I am worried about the space needed to store the results, and I wanted to start carefully...

But I now think that we should have "tree" variants of both bd_lvm_lvs and bd_lvm_lvinfo, and if you think that the space overhead is acceptable also for existing callers that are not interested in the sub-structure, then I would be very happy to change the existing bd_lvm_lvinfo and bd_lvm_lvs functions to always fill segs, data_lvs, and metadata_lvs.

(Earlier I thought that a tree variant of bd_lvm_lvinfo is not very useful since you likely need information for all sub-lvs immediately, but if you are only interested in a single lv, it is better to make the necessary multiple calls to bd_lvm_lvinfo to retrieve the whole tree below a single LV than it is to retrieve info for absolutely all LVs, which might be many many more.)

vojtechtrefny replied:
Ok, I guess we can do it this way. There's one other option -- we plan to do a new major release with API changes so we could change the API of `bd_lvm_lvs` and `bd_lvm_lvinfo`. But that's something we can decide when we're ready to release 3.0.

mvollmer commented:
I have added some tests and also implemented this for `lvm-dbus.c`.  However, lvm-cli and lvm-dbus  have some differences in behavior:

 - lvm-dbus does not expose lv_start_pe for segments, and always sets it to zero.
 - lvm-dbus completely omits segments that have lost their PV, instead of setting their pvdev to zero.

I think we can just adjust lvm-cli to behave like lvm-dbus does now.  Losing lv_start_pe does not affect our plans for Cockpit, and it is enough to know how many bytes are missing in total.  It's not necessary to know how many segments have lost their PV, or where on the PV those segments were.

mvollmer commented:
> I think we can just adjust lvm-cli to behave like lvm-dbus does now.

Done.

mvollmer commented:
@tbzatek, @vojtechtrefny, soooo, I think this is ready for a serious review.

vojtechtrefny reviewed (commented):
Looks good to me overall. I plan to do a more detailed review later this week and also try some manual testing with some more complex/crazy LVM setups.

vojtechtrefny commented:
The failures on 32bit Debian don't look random. `lvm_dbus_tests.LvmTestLVsMultiSegment` and `lvm_dbus_tests.LvmTestPartialLVs` fail both on stable and testing. It's probably a problem with lvmdbusd, but we should be sure before we skip the tests on 32bit.

vojtechtrefny commented on src/plugins/lvm-dbus.c:
```diff
@@ -1160,6 +1160,116 @@ static gchar* _lvm_metadata_lv_name (const gchar *vg_name, const gchar *lv_name,
+    while (g_variant_iter_next (&iter, "(&o@a(tts))", &pv, &pv_segs)) {
+      pv_name_prop = get_object_property (pv, PV_INTF, "Name", NULL);
+      if (pv_name_prop) {
+        g_variant_get (pv_name_prop, "&s", &pv_name);
+        g_variant_iter_init (&iter2, pv_segs);
+        while (g_variant_iter_next (&iter2, "(tt&s)", &pv_first_pe, &pv_last_pe, NULL)) {
+          BDLVMSEGdata *seg = g_new0(BDLVMSEGdata, 1);
+          seg->pv_start_pe = pv_first_pe;
+          seg->size_pe = pv_last_pe - pv_first_pe + 1;
+          seg->pvdev = g_strdup (pv_name);
```
> The failures on 32bit Debian don't look random. `lvm_dbus_tests.LvmTestLVsMultiSegment` and `lvm_dbus_tests.LvmTestPartialLVs` fail both on stable and testing. It's probably a problem with lvmdbusd, but we should be sure before we skip the tests on 32bit.

So lvmdbusd works correctly, seems like it's glib that doesn't like our code: `pv_name` is suddently `NULL` in the nested while here. I'm not sure why, we get only reference from `pv_name_prop` but it shouldn't suddenly be freed and it didn't went out of scope, right?

mvollmer replied:
The bug was to use `gsize` for `pv_first_pe` and `pv_last_pe`. It's only 4 bytes on 32 bit arches, but `g_variant_iter_next` will write 8 bytes for a "t" field. That was overwriting `pv_name` on the stack. I have pushed a fix.

tbzatek replied:
Awesome! Haven't really noticed this.

vojtechtrefny commented on src/lib/plugin_apis/lvm.api:
```diff
@@ -270,6 +270,91 @@ GType bd_lvm_vgdata_get_type () {
+}
+
+static BDLVMSEGdata **copy_segs (BDLVMSEGdata **segs) {
+    int len;
```
Please use glib types here, looks like GIR scanner doesn't like int on 32bit:

```
/usr/lib/gcc/i686-linux-gnu/10/include/stddef.h:424: syntax error, unexpected identifier in '  __float128 __max_align_f128 __attribute__((__aligned__(__alignof(__float128))));' at '__float128'
/home/jenkins/workspace/libblockdev-pr/arch/i686/distro/debian_11/type/libblockdev/src/lib/plugin_apis/lvm.h:244: syntax error, unexpected BASIC_TYPE in '    for (int i = 0; i < len; i++)' at 'int'
/home/jenkins/workspace/libblockdev-pr/arch/i686/distro/debian_11/type/libblockdev/src/lib/plugin_apis/lvm.h:244: syntax error, unexpected identifier in '    for (int i = 0; i < len; i++)' at 'i'
/home/jenkins/workspace/libblockdev-pr/arch/i686/distro/debian_11/type/libblockdev/src/lib/plugin_apis/lvm.h:254: syntax error, unexpected BASIC_TYPE in '    for (int i = 0; segs[i]; i++)' at 'int'
/home/jenkins/workspace/libblockdev-pr/arch/i686/distro/debian_11/type/libblockdev/src/lib/plugin_apis/lvm.h:254: syntax error, unexpected identifier in '    for (int i = 0; segs[i]; i++)' at 'segs'
```

mvollmer replied:
Yes, seems to be have been `for (int i)`. Fixed.

vojtechtrefny commented on src/plugins/lvm.c:
```diff
@@ -1974,6 +2128,52 @@ BDLVMLVdata* bd_lvm_lvinfo (const gchar *vg_name, const gchar *lv_name, GError *
+        if (table && (num_items == 19)) {
+          BDLVMLVdata *lvdata = get_lv_data_from_table (table, TRUE);
+          if (result)
+            merge_lv_data (result, lvdata);
```
Leaking `lvdata`.

mvollmer replied:
Fixed!

vojtechtrefny reviewed (approved):
Looks good to me. Other than the few notes and the weird 32bit issue.

tbzatek reviewed (approved):
Generally looks good, the multiple `GError` overwrites would need fixing.

mvollmer commented:
Thanks for the review!

Are you happy with the code duplication between `bd_lvm_lvinfo_tree` and `bd_lvm_lvinfo`, for example?  I am thinking we could refactor that a bit to not duplicate so much.

mvollmer commented:
(Rebased)

mvollmer commented on src/plugins/lvm-dbus.c:
```diff
@@ -1160,6 +1160,116 @@ static gchar* _lvm_metadata_lv_name (const gchar *vg_name, const gchar *lv_name,
+    if (!prop)
+        return NULL;
+
+    // Count number of segments
```
Done.

mvollmer commented on tests/lvm_dbus_tests.py:
```diff
@@ -1162,6 +1163,61 @@ def test_lvs(self):
+        self.assertEqual(info.segs[1].pvdev, self.loop_dev)
+        self.assertNotEqual(info.segs[0].pv_start_pe, info.segs[1].pv_start_pe)
+
+        lvs = BlockDev.lvm_lvs("testVG")
```
lvm_lvs_tree is now used in LvmTestPartialLVs.test_lvpartial.  Is that enough?

vojtechtrefny replied:
I didn't notice that, that's enought, thanks.

mvollmer commented on tests/lvm_dbus_tests.py:
```diff
@@ -1162,6 +1163,61 @@ def test_lvs(self):
+
+        lvs = BlockDev.lvm_lvs("testVG")
+        self.assertEqual(len(lvs), 2)
+        self.assertListEqual([lv.lv_name for lv in lvs], ["testLV", "testLV2"])
```
Done.

mvollmer commented on src/lib/plugin_apis/lvm.api:
```diff
@@ -270,6 +270,91 @@ GType bd_lvm_vgdata_get_type () {
+
+    for (int i = 0; segs[i]; i++)
+        bd_lvm_segdata_free (segs[i]);
+    (g_free) (segs);
```
There is a syntax error without these parentheses:

```
In file included from blockdev.c:6:
plugin_apis/lvm.h: In function ‘free_segs’:
plugin_apis/lvm.h:259:1: error: unknown type name ‘g_fre’
  259 | g_fre e (segs);
      | ^~~~~
plugin_apis/lvm.h:259:1: warning: parameter names (without types) in function declaration
```

tbzatek replied:
omg, just another fail of the `boilerplate_generator` :dizzy_face: :hankey:

mvollmer commented on src/plugins/lvm-dbus.c:
```diff
@@ -1160,6 +1160,116 @@ static gchar* _lvm_metadata_lv_name (const gchar *vg_name, const gchar *lv_name,
+    // Count number of segments
+    n_segs = 0;
+    g_variant_iter_init (&iter, prop);
+    while (g_variant_iter_loop (&iter, "(&o@a(tts))", NULL, &pv_segs))
```
I have changed the first loop to also use `g_variant_iter_next()`, which I think is the easier of the two to understand.

mvollmer commented on src/plugins/lvm-dbus.c:
```diff
@@ -1160,6 +1160,116 @@ static gchar* _lvm_metadata_lv_name (const gchar *vg_name, const gchar *lv_name,
+            if (g_strcmp0 (role, "image") == 0)
+              data_lvs[i_data++] = sublv_name;
+            else if (g_strcmp0 (role, "metadata") == 0)
+              metadata_lvs[i_metadata++] = sublv_name;
```
There are other possibilities, indeed. Good catch! Fixed by breaking out of the loop as soon as we find either "data" or "metadata", and freeing sublv_name if it was still unused.

mvollmer commented on src/plugins/lvm-dbus.c:
```diff
@@ -2645,6 +2759,32 @@ BDLVMLVdata* bd_lvm_lvinfo (const gchar *vg_name, const gchar *lv_name, GError *
+        /* the error is already populated */
+        return NULL;
+
+    ret = get_lv_data_from_props (props, error);
```
Yeah.. This is a copy of `bd_lvm_lvinfo`, so I changed both.

mvollmer commented on src/plugins/lvm-dbus.c:
```diff
@@ -2645,6 +2759,32 @@ BDLVMLVdata* bd_lvm_lvinfo (const gchar *vg_name, const gchar *lv_name, GError *
+    }
+    if (ret) {
+        ret->segs = _lvm_segs (vg_name, lv_name, error);
+        _lvm_data_and_metadata_lvs (vg_name, lv_name, &ret->data_lvs, &ret->metadata_lvs, error);
```
This pattern is copied from `bd_lvm_lvinfo`.  I agree it is questionable.  I think `error` should only bet set when the function returns `NULL`, no?  What should we do when, for example, _lvm_data_lv_name fails?  Should the whole function fail?

tbzatek replied:
Also moved to #769.

mvollmer commented on src/plugins/lvm-dbus.c:
```diff
@@ -2842,6 +2982,140 @@ BDLVMLVdata** bd_lvm_lvs (const gchar *vg_name, GError **error) {
+        } else if ((g_strcmp0 (ret[j]->segtype, "thin-pool") == 0) ||
+                   (g_strcmp0 (ret[j]->segtype, "cache-pool") == 0)) {
+            ret[j]->data_lv = _lvm_data_lv_name (ret[j]->vg_name, ret[j]->lv_name, &l_error);
+            ret[j]->metadata_lv = _lvm_metadata_lv_name (ret[j]->vg_name, ret[j]->lv_name, &l_error);
```
Same as above, code is copied from `bd_lvm_lvs`.

tbzatek replied:
I see. Opened #769 as a TODO, will fix separately.

mvollmer commented on tests/lvm_dbus_tests.py:
```diff
@@ -1162,6 +1163,61 @@ def test_lvs(self):
+
+        LvmPVVGLVTestCase._clean_up(self)
+
+    def test_lvs(self):
```
Done.

mvollmer commented:
Looks like kbd_test.KbdTestBcacheAttachDetach.test_bcache_attach_detach_full_path  is the only test that is failing, maybe we can get that to pass as well? :)

tbzatek commented:
> Looks like kbd_test.KbdTestBcacheAttachDetach.test_bcache_attach_detach_full_path is the only test that is failing, maybe we can get that to pass as well? :)

It's scheduled for slaughter in #763

mvollmer commented:
> > Looks like kbd_test.KbdTestBcacheAttachDetach.test_bcache_attach_detach_full_path is the only test that is failing, maybe we can get that to pass as well? :)
> 
> It's scheduled for slaughter in #763

:+1:

mvollmer commented:
Is this now blocked on #769?

vojtechtrefny reviewed (approved):
Thank you for the fixes, looks good!

vojtechtrefny commented:
> Is this now blocked on #769?

@tbzatek mentioned he wants to do a final review, but other than this is ready to be merged. Thank you.

mvollmer commented:
> > 
> 
> @tbzatek mentioned he wants to do a final review, but other than this is ready to be merged. Thank you.

Great!  Do you want me to squash the "FIXUP" commits?

tbzatek reviewed (approved):
Looks good, ship it!

> > @tbzatek mentioned he wants to do a final review, but other than this is ready to be merged. Thank you.
> 
> Great! Do you want me to squash the "FIXUP" commits?

Yes please, squash the `FIXUP` commits for a final merge.

mvollmer commented:
> Yes please, squash the `FIXUP` commits for a final merge.

Done, thanks!
